crates/distribution-types/src/uv_requirement.rs

```diff
@@ -0,0 +1,177 @@
+use std::fmt::{Display, Formatter};
+use std::path::PathBuf;
+
+use indexmap::IndexMap;
+use url::Url;
+
+use pep440_rs::VersionSpecifiers;
+use pep508_rs::{MarkerEnvironment, MarkerTree, VerbatimUrl, VersionOrUrl};
+use uv_git::GitReference;
+use uv_normalize::{ExtraName, PackageName};
+
+use crate::{ParsedUrl, ParsedUrlError};
+
+/// The requirements of a distribution, an extension over PEP 508's requirements.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Requirements {
+    pub dependencies: Vec<Requirement>,
+    pub optional_dependencies: IndexMap<ExtraName, Vec<Requirement>>,
+}
+
+/// A representation of dependency on a package, an extension over a PEP 508's requirement.
+#[derive(Hash, Debug, Clone, Eq, PartialEq)]
+pub struct Requirement {
+    pub name: PackageName,
+    pub extras: Vec<ExtraName>,
+    pub marker: Option<MarkerTree>,
+    pub source: RequirementSource,
+}
+
+impl Requirement {
+    /// Returns whether the markers apply for the given environment.
+    pub fn evaluate_markers(&self, env: &MarkerEnvironment, extras: &[ExtraName]) -> bool {
+        if let Some(marker) = &self.marker {
+            marker.evaluate(env, extras)
+        } else {
+            true
+        }
+    }
+
+    pub fn from_requirement(requirement: pep508_rs::Requirement) -> Result<Self, ParsedUrlError> {
+        let source = match requirement.version_or_url {
+            None => RequirementSource::Registry {
+                version: VersionSpecifiers::empty(),
+                index: None,
+            },
+            // The most popular case: Just a name, a version range and maybe extras.
+            Some(VersionOrUrl::VersionSpecifier(version)) => RequirementSource::Registry {
+                version,
+                index: None,
+            },
+            Some(VersionOrUrl::Url(url)) => {
+                let direct_url = ParsedUrl::try_from(&url.to_url())?;
```

Comment:
Why does this take a reference and not an owned value? Can't we use the value directly for `ParsedLocalFileUrl` at least?